nebari-dev/nebari
# Refactor role creation for upgrade command path

## Reference Issues or PRs
closes #2766 



## What does this implement/fix?

This issue is a patching fix for the upgrade command present in the previous release, currently the upgrade logic when requesting the user to perform the role creation (more details see linked issue), assumes the presence of the role when assigning it to the legacy groups. However, this leads to errors when the role does not exist or is within Terraform if the user attempts to manually address the missing role to continue the upgrade.

This PRs includes a new section in the previous code logic to create the role, and to avoid conflicts with terraform, I adopted a "legacy" prefix to the role name with a befitting description for future reference when the amdins manages keycloak in the future.  

_Put a `x` in the boxes that apply_


## Testing


## How to test this PR?

- Perform any type of deployment for Nebari `v2024.7.1`, `local` is advised, 
- create some testing groups and users with such users assigned to some of these groups (one user and one group should suffice). 
- Launch one user instance and assert the folders are mounted;
- Install this nebari version in dev mode and perform the upgrade; (You will need to create a tag  locally for proper installation of nebari)
- Assert the new role has been created in the jupyterhub client, and the previous groups all have the role assignment;
- Perform the deployment and launch a user instance to validate the folder mounting is kept the same (consider logging out first to avoid issues with caching of user_Data)



## Any other comments?
